### PR TITLE
feat: support systemd docker cgroup driver

### LINKED ISSUE #363: Generalize docker cgroup driver support
Currently the agent adds our own cgroup control routines to retrieve container statistics of containers after their termination on top of Docker's cgroupfs lifecycle.

This works well with the default cgroup driver, but there are reports that it conflicts with the "systemd" cgroup driver.  The relevant configuration option name is `native.cgroupdriver`.

There seems no big technical difference between systemd-managed cgroups and cgroupfs-based cgroups, but they have different directory namings under the sysfs.  It is also notable that Kubernetes recommends to use the "systemd" cgroup driver and keep the configurations for kubelets and Docker same.

* https://stackoverflow.com/questions/43794169/docker-change-cgroup-driver-to-systemd/65870152
* https://kubernetes.io/docs/setup/production-environment/container-runtimes/
* https://kubernetes.io/docs/tasks/administer-cluster/kubeadm/configure-cgroup-driver/

Let's support both types of cgroup drivers to prevent unexpected conflicts on various systems, by adding auto-detection of the current cgroup driver and using different directory names that match with each driver when collecting stats.

### PR DESCRIPTION
**Pull request by qkoo0833**
1. Automatically detect docker's current cgroup driver

2. Match different path prefix with each driver

resolve #363

### CODE REVIEW

**Review by achimnol — CHANGES_REQUESTED** [review]
Let's use [the new pattern matching syntax](https://peps.python.org/pep-0636/)!
This will allow type checkers to validate whether all enum branches are explicitly taken or not.

File: src/ai/backend/agent/stats.py [hunk 1]
```diff
@@ -266,7 +281,7 @@ def __init__(
         self, agent: "AbstractAgent", mode: StatModes = None, *, cache_lifespan: int = 120
     ) -> None:
         self.agent = agent
-        self.mode = mode if mode is not None else StatModes.get_preferred_mode()
+        self.mode = mode if mode == StatModes.DOCKER else StatModes.get_preferred_mode()
```
Comment (achimnol) [inline src/ai/backend/agent/stats.py]: The argument is optional and `None` means that the caller is leaving the choice of mode to the implementation. So, this should not be changed.

File: src/ai/backend/agent/stats.py [hunk 2]
```diff
@@ -66,7 +68,9 @@ def check_cgroup_available():
 
 class StatModes(enum.Enum):
     CGROUP = "cgroup"
+    CGROUPFS = "cgroupfs"
     DOCKER = "docker"
+    SYSTEMD = "systemd"
 
     @staticmethod
     def get_preferred_mode():
```
Comment (achimnol) [inline src/ai/backend/agent/stats.py]: This function should be updated to auto-detect the cgroupfs/systemd modes as well.

**Review by achimnol — CHANGES_REQUESTED** [review]

File: src/ai/backend/agent/stats.py [hunk 3]
```diff
@@ -65,18 +68,39 @@ def check_cgroup_available():
 
 
 class StatModes(enum.Enum):
+    """
+    CGROUP = Tries to automatically detect docker's current cgroup driver
+    CGROUPFS = Manually set cgroup driver to cgroupfs
+    SYSTEMD = Manually set cgroup driver to systemd
```
Comment (achimnol) [inline src/ai/backend/agent/stats.py:75]: Please give a small favor to `DOCKER` as well. :wink:

File: src/ai/backend/agent/stats.py [hunk 4]
```diff
@@ -65,18 +68,39 @@ def check_cgroup_available():
...
-            return StatModes.CGROUP
+            docker_cgroup_driver = await StatModes.check_docker_cgroup_driver()
+            if docker_cgroup_driver is not None:
+                return docker_cgroup_driver
+            else:  # Fallback to default (cgroupfs)
+                return StatModes.CGROUPFS
         return StatModes.DOCKER
 
+    @classmethod
+    async def check_docker_cgroup_driver(cls):
```
Comment (achimnol) [inline src/ai/backend/agent/stats.py]: Could you specify the return type explicitly?
If required, you may add `from __future__ import annotations`.

**Review by achimnol — CHANGES_REQUESTED** [review]

File: src/ai/backend/agent/stats.py [hunk 5]
```diff
@@ -93,7 +95,7 @@ async def get_preferred_mode():
         return StatModes.DOCKER
 
     @classmethod
-    async def check_docker_cgroup_driver(cls):
+    async def check_docker_cgroup_driver(cls) -> StatModes:
```
Comment (achimnol) [inline src/ai/backend/agent/stats.py]: I think this should be `Optional` as the method may return `None`.

Comment (kmkwon94) [reply src/ai/backend/agent/stats.py]: Hmm? I alreday applied Optional type in this commit 44bb4cf

Comment (achimnol) [reply src/ai/backend/agent/stats.py]: Ah, my mistake. I just checked the first commit only; Thanks!

File: src/ai/backend/agent/stats.py [hunk 6]
```diff
@@ -65,18 +69,40 @@ def check_cgroup_available():
...
+                return docker_cgroup_driver
+            else:  # Fallback to default (cgroupfs)
+                return StatModes.CGROUPFS
         return StatModes.DOCKER
 
+    @classmethod
+    async def check_docker_cgroup_driver(cls) -> Optional[StatModes]:
+        async with Docker() as docker:
+            try:
+                result = await docker._query_json("info", method="GET")
```
Comment (sanxiyn) [inline src/ai/backend/agent/stats.py:101]: This should be `docker.system.info()`.

**Comment by achimnol** [comment]
This is replaced by #1000 with better handling of 4 combinations of (systemd, cgroup) x (cgroup v1, cgroup v2).

**Pull request closed without merging**